include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -255,6 +272,205 @@ class TensorProductMatrixSymmetricSum
+                {
+                  a += std::abs(M_0[i][j][v]);
+                  b += std::abs(M_1[i][j][v]);
+                }
+
+            mask[v] = (a != 0.0) && (b != 0.0);
+          }
+
+        const auto eps = std::pow<NumberScalar>(
+          static_cast<NumberScalar>(10.0),
+          static_cast<NumberScalar>(
+            std::log10(std::numeric_limits<NumberScalar>::epsilon()) / 2.0));
+
+        const auto less = [eps](const auto a, const auto b) -> bool {
+          return (b - a) > std::max(eps, std::abs(a + b) * eps);
+        };
+
+        const auto greater = [eps](const auto a, const auto b) -> bool {
+          return (a - b) > std::max(eps, std::abs(a + b) * eps);
+        };
+
+        for (unsigned int v = 0; v < n_lanes; ++v)
+          if (mask[v])
+            for (unsigned int i = 0; i < M_0.size(0); ++i)
+              for (unsigned int j = 0; j < M_0.size(1); ++j)
+                if (less(M_0[i][j][v], M_1[i][j][v]))
+                  return true;
+                else if (greater(M_0[i][j][v], M_1[i][j][v]))
+                  return false;
+
+        for (unsigned int v = 0; v < n_lanes; ++v)
+          if (mask[v])
+            for (unsigned int i = 0; i < K_0.size(0); ++i)
+              for (unsigned int j = 0; j < K_0.size(1); ++j)
+                if (less(K_0[i][j][v], K_1[i][j][v]))
+                  return true;
+                else if (greater(K_0[i][j][v], K_1[i][j][v]))
+                  return false;
+
+        return false;
+      }
+    };
+  } // namespace TensorProductMatrixSymmetricSum
+} // namespace internal
+
+
+
+/**
+ * A class similar to TensorProductMatrixSymmetricSum.
+ *
+ * The class TensorProductMatrixSymmetricSum stores a
+ * 1D mass matrix, 1D stiffness matrix, eigenvalues and eigenvectors
+ * for each direction. If one uses one TensorProductMatrixSymmetricSum
+ * instance for, e.g., each cell, these quantities are stored
+ * for each cell. There is no possibility to reusage quanties between
+ * TensorProductMatrixSymmetricSum instances even if the values of the
+ * internal data structures might be the same. This class targets the case
+ * that one needs many TensorProductMatrixSymmetricSum instances and
+ * there might be the possibility to reuse the internal quanties.
+ *
+ * This class is flexible and allows to interpret the parameter
+ * @p index arbitraryly. In the case of an element-centric patch
+ * smoother, the index might correspond to the cell index and,
+ * in the case of a vertex patch smoother, the index might
+ * correspond to the vertex index defining the vertex patch. If
+ * @p n_rows_1d is set to -1, the sizes of the mass matrices and
+ * the stiffness matrices can differ between cells, which might be
+ * useful if the class is used in the context of hp-refinement to
+ * construct a smoother.
+ */
+template <int dim, typename Number, int n_rows_1d = -1>
+class TensorProductMatrixSymmetricSumCache
```

Comment:
Maybe someone has a better name for the class?

Reply:
Regarding the name, I think `Cache` is not really the right name, because to me it would indicate that `TensorProductMatrixSymmetricSum` would compute something on the fly that gets pre-computed in the present class. However, what this class actually represents is a collection of many `TensorProductMatrixSymmetricSum` objects, possibly re-using some of the internally stored objects. So I would vote for `...Collection` or `...Storage` - the fact that we indeed compress the internal representation is not the main information a user would need to have and thus be reflected in the name, but it is rather the fact that we collect many objects.

Reply:
I'll go with `TensorProductMatrixSymmetricSumCollection`!